Fix incorrect transforms on some exported Revit elements

corybarr opened the pull request:
A small percentage of Revit elements in some datasets were being exported with incorrect transformation stacks.

To simplify the logic, this PR flattens the glTF node hierarchy. Every exported Revit element is a direct child of the `xFormNode` in the exported glTF:

 `rootNode -> xFormNode -> EVERYTHING_ELSE`

with every element now having a correspondingly simply transform stack.

lilleyse commented:
The problem with flattening the hierarchy is that parent/child relationships get lost. This doesn't really matter today with how we've configured the design tiler, but it could matter in the future.

corybarr commented:
> The problem with flattening the hierarchy is that parent/child relationships get lost. This doesn't really matter today with how we've configured the design tiler, but it could matter in the future.

Yes, that's true. We could add it back now. In theory, this is a very simple operation. I'm not 100% sure why the nested hierarchy had occasional issues. It would need investigation. (hunch: maybe in some instances Revit exports children before parents and the logic isn't handling that correctly.)

> could matter in the future.

I'm assuming you're thinking of how to leverage BIM structure to boost streaming performance. There are a few ways to convert BIM relations into parent-child structures. It might be more efficient to define how to build the hierarchy when we're about to leverage it. Right now we only use one reasonable parent-child relationship: It's a child if and only if `familyInstance.SuperComponent != null` 

In the future, we probably want to also add `category` and/or `host` in some combination. Quick example:

<img width="1063" height="1174" alt="image" src="https://github.com/user-attachments/assets/8bfc5d4a-a675-42a0-a6bf-a912f463bc39" />

The window is made of multiple elements. The main window is a `supercomponent` of the frame. We capture that. In addition, the wall is a `host` of the window. We don't capture that. Parenting elements to their `host` might make more performant tilesets.

Since it's rare that any single individual in their respective discipline needs to see all categories, parenting to the category could make tree traversal more performant by not traversing what people don't need:

- root
  - xForm
    - Walls
      - wall1
      - ...
    - Windows
      - window1
        - windowPart
      - ...   

Because of the options, we might have more insight into how to construct the hierarchy while implementing the task that needs it. But again, it should be quite easy to add back the `supercomponent` parent-child relationship. Doing this as a post-processing step should eliminate all errors and set us up for more advanced tree structures. I do wonder if it's a bit of pre-mature optimization though.

> parent/child relationships get lost.

Technically, this is still captured if metadata is exported. 

What do you think? Keep flattened until we use it, or add back `supercomponent` parenting now?

corybarr commented:
I added `SuperComponent` parenting back. I've tested this patch on a previously troublesome large client model and Snowdon. I'm not able to break it.

I also cleaned and refactored a slight amount. Adding more complex hierarchies should be easier.

The fundamental issue is that the Revit API is set up to make things easy for you if you always bake the current transform stack into the vertices (and normals). This is done during `OnPolymesh`. It's tricky if you don't.

It means reverse engineering some variations on the transform stack that can wrap `OnPolyMesh`. The solution to this particular issue was correctly handling when an element's parent did not make use of it's Revit transform. It happens when the event stack does this:

```
- OnElementBegin
  - OnInstanceBegin (push Revit transform stack)
  - OnInstanceEnd (pop Revit transform stack)
  - OnPolymesh (Revit's element transform is never used)
- OnElementEnd
```
Usually elements work like this:

```
- OnElementBegin
  - OnInstanceBegin (push Revit transform stack)
    - OnPolymesh (use Revit's transform stack)
  - OnInstanceEnd (pop Revit transform stack)
- OnElementEnd
```

r-veenstra commented:
This looks okay and my test exports work for me.  I'm obviously not deep in the details of Revit's export process but what you've explained makes sense.

r-veenstra reviewed (approved):
LGTM
